### ISSUE TITLE
Assertion failure in index_set.cpp:682: !more_before || index >= std::prev(it)->second

### ISSUE THREAD

**Issue by jpsim** [realm/realm-cocoa#3537]
Here's a report from a user which occurs when preparing the change information for notification blocks with 0.100 or later. Does not reproduce with 0.99.1.

```
Realm/ObjectStore/index_set.cpp:682: [realm-core-0.100.0] Assertion failed: !more_before || index >= std::prev(it)->second
0 Realm 0x00000001099ee248 _ZN5realm4util18terminate_internalERNSt3__118basic_stringstreamIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 40
1 Realm 0x00000001099ee514 _ZN5realm4util9terminateEPKcS2_l + 452
2 Realm 0x00000001097c5d8b _ZN5realm8IndexSet6do_addENS_5_impl33MutableChunkedRangeVectorIteratorINSt3__111__wrap_iterIPNS1_18ChunkedRangeVector5ChunkEEEEEm + 651
3 Realm 0x00000001097c67e8 _ZN5realm8IndexSet11add_shiftedEm + 488
4 Realm 0x0000000109777b30 _ZN5realm5_impl23CollectionChangeBuilder9move_overEmmb + 1072
5 Realm 0x00000001099ce892 _ZN12_GLOBAL__N_116LinkViewObserver10erase_rowsEmmmb + 1330
6 Realm 0x00000001099ccf30 _ZN5realm5_impl17TransactLogParser9parse_oneIN12_GLOBAL__N_116LinkViewObserverEEEvRT_ + 2864
7 Realm 0x00000001099cc027 _ZN5realm5_impl17TransactLogParser5parseIN12_GLOBAL__N_116LinkViewObserverEEEvRNS0_17NoCopyInputStreamERT_ + 87
8 Realm 0x00000001099cbdf6 _ZN5realm11SharedGroup15do_advance_readIN12_GLOBAL__N_116LinkViewObserverEEEbPT_NS0_9VersionIDERNS_5_impl7HistoryE + 486
9 Realm 0x00000001099cbbfc _ZN5realm11SharedGroup12advance_readIN12_GLOBAL__N_116LinkViewObserverEEEvPT_NS0_9VersionIDE + 396
10 Realm 0x00000001099cba5b _ZN5realm5_impl17SharedGroupFriend12advance_readIN12_GLOBAL__N_116LinkViewObserverEEEvRNS_11SharedGroupEPT_NS5_9VersionIDE + 59
11 Realm 0x00000001099c201b _ZN5realm14LangBindHelper12advance_readIN12_GLOBAL__N_116LinkViewObserverEEEvRNS_11SharedGroupEOT_NS4_9VersionIDE + 59
12 Realm 0x00000001099c1f8f _ZN5realm5_impl11transaction7advanceERNS_11SharedGroupERNS0_21TransactionChangeInfoENS2_9VersionIDE + 191
13 Realm 0x000000010980d927 _ZN12_GLOBAL__N_121IncrementalChangeInfo16advance_to_finalEN5realm11SharedGroup9VersionIDE + 183
14 Realm 0x000000010980bd02 _ZN5realm5_impl16RealmCoordinator19run_async_notifiersEv + 6210
15 Realm 0x000000010980a245 _ZN5realm5_impl16RealmCoordinator9on_changeEv + 37
16 Realm 0x00000001097be63f _ZN5realm5_impl20ExternalCommitHelper6listenEv + 735
17 Realm 0x00000001097bf3d7 _ZZN5realm5_impl20ExternalCommitHelperC1ERNS0_16RealmCoordinatorEENK3$_0clEv + 23
18 Realm 0x00000001097bf3ad _ZNSt3__112__async_funcIZN5realm5_impl20ExternalCommitHelperC1ERNS2_16RealmCoordinatorEE3$_0JEE9__executeIJEEEvNS_15__tuple_indicesIJXspT_EEEE + 61
19 Realm 0x00000001097bf365 _ZNSt3__112__async_funcIZN5realm5_impl20ExternalCommitHelperC1ERNS2_16RealmCoordinatorEE3$_0JEEclEv + 21
20 Realm 0x00000001097bf28f _ZNSt3__119__async_assoc_stateIvNS_12__async_funcIZN5realm5_impl20ExternalCommitHelperC1ERNS3_16RealmCoordinatorEE3$_0JEEEE9__executeEv + 31
21 Realm 0x00000001097bfd04 _ZNSt3__114__thread_proxyINS_5tupleIJMNS_19__async_assoc_stateIvNS_12__async_funcIZN5realm5_impl20ExternalCommitHelperC1ERNS5_16RealmCoordinatorEE3$_0JEEEEEFvvEPSB_EEEEEPvSG_ + 532
22 libsystem_pthread.dylib 0x000000010fb6799d _pthread_body + 131
23 libsystem_pthread.dylib 0x000000010fb6791a _pthread_body + 0
24 libsystem_pthread.dylib 0x000000010fb65351 thread_start + 13
```

This only occurs with a results in `Mode::Table` mode.

**Comment by bdash** [comment]
#3516 claims to have fixed this, and was present in v0.101. What version of Realm was this seen with?

**Comment by austinzheng** [comment]
From the support ticket, both 0.100 and 0.101.

**Comment by tgoyne** [comment]
While it's the same assert failing, the stack trace here is different from what was fixed in #3516.
